lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -1747,166 +1761,188 @@ export class UserAgentApplication {
+                this.user = response.account = null;
                 // TODO: optimize this - may be combine if it is a string in both cases
-                this.cacheStorage.setItem(Constants.loginError, "Nonce Mismatch. Expected Nonce: " + this.cacheStorage.getItem(Constants.nonceIdToken, this.inCookie) + "," + "Actual Nonce: " + idToken.nonce);
-                this.logger.error("Nonce Mismatch.Expected Nonce: " + this.cacheStorage.getItem(Constants.nonceIdToken, this.inCookie) + "," + "Actual Nonce: " + idToken.nonce);
+                this.cacheStorage.setItem(Constants.loginError, "Nonce Mismatch. Expected Nonce: " + this.cacheStorage.getItem(Constants.nonceIdToken, this.inCookie) + "," + "Actual Nonce: " + response.idToken.nonce);
+                this.logger.error("Nonce Mismatch.Expected Nonce: " + this.cacheStorage.getItem(Constants.nonceIdToken, this.inCookie) + "," + "Actual Nonce: " + response.idToken.nonce);
+                error = ClientAuthError.createNonceMismatchError(this.cacheStorage.getItem(Constants.nonceIdToken, this.inCookie), response.idToken.nonce);
               }
               // Save the token
               else {
-                this.cacheStorage.setItem(Constants.idTokenKey, tokenResponse.parameters[Constants.idToken]);
+                this.cacheStorage.setItem(Constants.idTokenKey, parameters[Constants.idToken]);
                 this.cacheStorage.setItem(Constants.msalClientInfo, clientInfo);
 
                 // Save idToken as access token for app itself
-                this.saveAccessToken(authority, tokenResponse, this.user, clientInfo, idToken);
+                this.saveAccessToken(response, authority, parameters, clientInfo);
               }
             } else {
               // TODO: avoid repeated strings - will this be optimized with error handling?
-              authorityKey = tokenResponse.stateResponse;
-              acquireTokenUserKey = tokenResponse.stateResponse;
+              authorityKey = stateInfo.state;
+              acquireTokenUserKey = stateInfo.state;
               this.logger.error("Invalid id_token received in the response");
-              tokenResponse.parameters["error"] = "invalid idToken";
-              tokenResponse.parameters["error_description"] = "Invalid idToken. idToken: " + tokenResponse.parameters[Constants.idToken];
+              parameters["error"] = "invalid idToken";
+              parameters["error_description"] = "Invalid idToken. idToken: " + parameters[Constants.idToken];
               this.cacheStorage.setItem(Constants.msalError, "invalid idToken");
-              this.cacheStorage.setItem(Constants.msalErrorDescription, "Invalid idToken. idToken: " + tokenResponse.parameters[Constants.idToken]);
+              this.cacheStorage.setItem(Constants.msalErrorDescription, "Invalid idToken. idToken: " + parameters[Constants.idToken]);
+              error = ClientAuthError.createInvalidIdTokenError(response.idToken);
```

Comment:
just match the error obj in cache to the obj returned

Reply:
done